frame/assets-freezer/src/lib.rs

```diff
@@ -0,0 +1,361 @@
+		pub(super) reserved: Balance,
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The fungibles trait impl whose assets this reserves.
+		type Assets: fungibles::InspectWithoutFreezer<Self::AccountId>;
+
+		/// Place to store the fast-access freeze data for the given asset/account.
+		type Store: StoredMap<(AssetIdOf<Self>, Self::AccountId), FreezeData<BalanceOf<Self>>>;
+	}
+
+	//
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T> = "Balance", AssetIdOf<T> = "AssetId")]
+	pub enum Event<T: Config> {
+		/// An asset has been reserved.
+		/// \[asset, who, amount\]
+		Held(AssetIdOf<T>, T::AccountId, BalanceOf<T>),
+		/// An asset has been unreserved.
+		/// \[asset, who, amount\]
+		Released(AssetIdOf<T>, T::AccountId, BalanceOf<T>),
+	}
+
+	// No new errors
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	// No hooks.
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	// No calls.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> FrozenBalance<AssetIdOf<T>, T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn frozen_balance(id: AssetIdOf<T>, who: &T::AccountId) -> Option<BalanceOf<T>> {
+		let f = T::Store::get(&(id, who.clone()));
+		if f.reserved.is_zero() { None } else { Some(f.reserved) }
+	}
+}
+
+impl<T: Config> fungibles::Inspect<<T as SystemConfig>::AccountId> for Pallet<T> {
+	type AssetId = AssetIdOf<T>;
+	type Balance = BalanceOf<T>;
+	fn total_issuance(asset: AssetIdOf<T>) -> BalanceOf<T> {
+		T::Assets::total_issuance(asset)
+	}
+	fn minimum_balance(asset: AssetIdOf<T>) -> BalanceOf<T> {
+		T::Assets::minimum_balance(asset)
+	}
+	fn balance(asset: AssetIdOf<T>, who: &T::AccountId) -> BalanceOf<T> {
+		T::Assets::balance(asset, who)
+	}
+	fn reducible_balance(asset: AssetIdOf<T>, who: &T::AccountId, keep_alive: bool) -> BalanceOf<T> {
+		T::Assets::reducible_balance(asset, who, keep_alive)
+	}
+	fn can_deposit(asset: AssetIdOf<T>, who: &T::AccountId, amount: BalanceOf<T>)
+		-> DepositConsequence
+	{
+		T::Assets::can_deposit(asset, who, amount)
+	}
+	fn can_withdraw(
+		asset: AssetIdOf<T>,
+		who: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> WithdrawConsequence<BalanceOf<T>> {
+		T::Assets::can_withdraw(asset, who, amount)
+	}
+}
+
+impl<T: Config> fungibles::Transfer<<T as SystemConfig>::AccountId> for Pallet<T> where
+	T::Assets: fungibles::Transfer<T::AccountId>
+{
+	fn transfer(
+		asset: Self::AssetId,
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: Self::Balance,
+		death: WhenDust,
+	) -> Result<Self::Balance, DispatchError> {
+		T::Assets::transfer(asset, source, dest, amount, death)
```

Comment:
Feeling a little stupid, but how is this right?

Shouldn't transfer check the freezer before delegating to the `Assets` implementation?